ethexe/compute/src/service.rs

```diff
@@ -156,62 +162,71 @@ impl<P: ProcessorExt> Stream for ComputeService<P> {
             }
         }
 
-        if matches!(self.blocks_state, State::WaitForBlock) {
+        if let State::WaitForBlock = &self.blocks_state {
             match self.blocks_queue.pop_back() {
                 Some(BlockAction::Prepare(block)) => {
-                    let PrepareInfo {
-                        chain,
-                        missing_codes,
-                        missing_validated_codes,
-                    } = prepare::prepare(&self.db, block)?;
-
-                    self.blocks_state = State::WaitForCodes {
-                        block,
-                        chain,
-                        waiting_codes: missing_validated_codes,
+                    let MissingData {
+                        codes,
+                        validated_codes,
+                    } = prepare::missing_data(&self.db, block)?;
+
+                    debug_assert!(
+                        validated_codes
+                            .iter()
+                            .all(|code_id| codes.contains(code_id)),
+                        "All missing validated codes must be in the missing codes list"
+                    );
+
+                    self.blocks_state = State::WaitForRequestedData {
+                        block_hash: block,
+                        codes: validated_codes,
                     };
 
-                    if !missing_codes.is_empty() {
-                        return Poll::Ready(Some(Ok(ComputeEvent::RequestLoadCodes(
-                            missing_codes,
-                        ))));
+                    if !codes.is_empty() {
+                        return Poll::Ready(Some(Ok(ComputeEvent::RequestLoadCodes(codes))));
                     }
                 }
-                Some(BlockAction::Process(block)) => {
-                    if !self.db.block_meta(block).prepared {
-                        return Poll::Ready(Some(Err(ComputeError::BlockNotPrepared(block))));
-                    }
-
-                    self.blocks_state = State::ComputeBlock(
-                        compute::compute(self.db.clone(), self.processor.clone(), block).boxed(),
-                    );
+                Some(BlockAction::Compute(announce)) => {
+                    self.blocks_state = State::Computation {
+                        announce_hash: announce.hash(),
+                        future: compute::compute(self.db.clone(), self.processor.clone(), announce)
+                            .boxed(),
+                    };
                 }
                 None => {}
             }
         }
 
-        if let State::WaitForCodes {
-            block,
-            chain,
-            waiting_codes,
-        } = &self.blocks_state
-            && waiting_codes.is_empty()
+        if let State::WaitForRequestedData { block_hash, codes } = &self.blocks_state
```

Comment:
```suggestion
        self.block_state =  match self.blocks_state { ... }
```
It should be one match instead of few if branches so state transition is clearly seen